beacon-chain/core/altair/transition.go

```diff
@@ -0,0 +1,110 @@
+package altair
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	e "github.com/prysmaticlabs/prysm/beacon-chain/core/epoch"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/epoch/precompute"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"go.opencensus.io/trace"
+)
+
+// ProcessEpoch describes the per epoch operations that are performed on the beacon state.
+// It's optimized by pre computing validator attested info and epoch total/attested balances upfront.
+//
+// Spec code:
+// def process_epoch(state: BeaconState) -> None:
+//    process_justification_and_finalization(state)  # [Modified in Altair]
+//    process_inactivity_updates(state)  # [New in Altair]
+//    process_rewards_and_penalties(state)  # [Modified in Altair]
+//    process_registry_updates(state)
+//    process_slashings(state)  # [Modified in Altair]
+//    process_eth1_data_reset(state)
+//    process_effective_balance_updates(state)
+//    process_slashings_reset(state)
+//    process_randao_mixes_reset(state)
+//    process_historical_roots_update(state)
+//    process_participation_flag_updates(state)  # [New in Altair]
+//    process_sync_committee_updates(state)  # [New in Altair]
+func ProcessEpoch(ctx context.Context, state state.BeaconStateAltair) (state.BeaconStateAltair, error) {
```

Comment:
Any chance we could shift to a functional options pattern here?

```go
state, err := withTransitionPipelines(
  headState,
  processEpochParticipation(bp, vp),
  processJustificationAndFinalizationPrecompute(bp),
  ...
)
```
maybe we can open a tracking issue for it, but it would remove all those if err checks

Reply:
Yes, it seems like a good first issue type of candidate